File: 86_partition_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        
        beforeHead = ListNode(0)
        before = beforeHead
        afterHead = ListNode(0)
        after = afterHead
        while(head):
            if head.val >= x:
                after.next = head
                after = after.next
            else:
                before.next = head
                before = before.next
            head = head.next
        
        after.next = None        
        before.next = afterHead.next
        
        return beforeHead.next

File: Set1_1000/86_partition_list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        if (head == null)
            return head;
        ListNode current = head;
        ListNode lessThanX = new ListNode(0);
        ListNode tempLessThanX = lessThanX;
        ListNode greaterThanOrEqualToX = new ListNode(0);
        ListNode tempGreaterThanOrEqualToX = greaterThanOrEqualToX;
        while (current != null) {
            if (current.val < x) {
                tempLessThanX.next = new ListNode(current.val);
                tempLessThanX = tempLessThanX.next;
            }
            else {
                tempGreaterThanOrEqualToX.next = new ListNode(current.val);
                tempGreaterThanOrEqualToX = tempGreaterThanOrEqualToX.next;
            }
            current = current.next;
        }
        tempLessThanX.next = greaterThanOrEqualToX.next;
        return lessThanX.next;
    }
}
